Implement NatsJetStreamChannel based on latest proposal spec

dan-j opened the pull request:
Fixes #145

## Proposed Changes

- 🎁  Cluster or namespace-scoped dispatcher
- 🎁  NATS credentials per dispatcher
- 🎁  Supports configuring stream and consumer options with sensible defaults
- 🧽  Refactored some directory structures to be more inline with eventing-kafka
- 🗑️  Remove original JetStream implementation

Outstanding things:

- Unit tests
- E2E tests
- DeliveryPolicy should default to New and not All, but this depends on adding a `JetStreamConsumerReady` condition

**Release Note**


```release-note
Alpha support for JetStream via the new `NatsJetStreamChannel` CRD type.
```

**Docs**

<!--
📖 If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.
-->

We should probably update the docs to include a reference to this channel implementation

dan-j commented:
Apologies.. it's a biggie

If you have any suggestions for me to make this easier to review let me know.

dan-j commented:
Was meant to add a bit more context when creating the PR but I was pushed for time...

This isn't ready for merge yet, I have only tested that I can create a channel using defaults in the knative-eventing namespace and everything is working. I also need to write tests which will take a while. I thought I'd create the PR so that people could get a view into what I've done so far and get some feedback.

I haven't tested:

- things are cleaned up when channels are deleted
- cluster-scoped channels created in other namespaces (managed by `knative-eventing/jetstream-ch-dispatcher`)
- namespace-scoped channels in other namespaces (managed by a `jetstream-ch-dispatcher` deployment in the same namespace
- custom settings on the channel `.spec.stream` and `.spec.consumerConfigTemplate`

dan-j commented:
If somebody wants to try this out (also needs adding to the docs):

```
ko apply -f config/webhook
ko apply -f config/jetstream
kubectl apply -f examples/config-nats.yaml
kubectl apply -f examples/NatsJetstreamChannel-simple.yaml
```

dan-j commented on docs/jetstream.md:
```diff
@@ -0,0 +1,214 @@
+- Secured NATS clusters with a Credentials File (https://docs.nats.io/developing-with-nats/security/creds)
+    - Globally-scoped credentials by a ConfigMap/Secret in the `knative-eventing` namespace
+    - Namespace-scoped credentials by a ConfigMap/Secret in the same namespace as the `NatsJetStreamChannel`
+    - Channel-scoped credentials specified by a Secret and referenced in the `NatsJetStreamChannel` resource.
```
This needs removing, we're only going to allow credentials scoped to a dispatcher (i.e. globally or per-namespace)

dan-j commented on docs/jetstream.md:
```diff
@@ -0,0 +1,214 @@
+    - Namespace-scoped credentials by a ConfigMap/Secret in the same namespace as the `NatsJetStreamChannel`
+    - Channel-scoped credentials specified by a Secret and referenced in the `NatsJetStreamChannel` resource.
+- TLS-enabled clusters (https://docs.nats.io/developing-with-nats/security/tls)
+    - CA should be defined in a ConfigMap/Secret in the `knative-eventing` namespace.
```
the `config-nats` configmap should tell the dispatcher the CA, either a base64'd string or a ref to a secret

dan-j commented on docs/jetstream.md:
```diff
@@ -0,0 +1,214 @@
+                                    # both 30 and 30% as valid values
+        maxAckPending: 0            # number of outstanding messages awaiting ack before suspending delivery, 0 denotes 
+                                    # no maximum
+        idleHeartbeat: ""           # if set, a time.Duration for the idle heartbeat the server sends to the client
```
`idleHeartbeat` needs removing, it's not allowed for queue subscribers

dan-j commented:
Haven't forgot about this, just had other stuff going on. I doubt I'm going to get much time on this until mid-December.

zhaojizhuang commented:
> Haven't forgot about this, just had other stuff going on. I doubt I'm going to get much time on this until mid-December.

@dan-j Great !

dan-j commented:
Picking this back up now, I expect it will still be a few weeks before finishing as I won't be full time on it, but will keep everyone posted.

dan-j commented:
Few small updates to this, didn't realise how long it had been since I started this work, took a while to familiarise myself with it again.

Adding some docs on running everything locally in a k3d cluster, much of my previous list of outstanding things is still to do, but I should be spending more time to get this in a production-ready state in the new few weeks

dan-j commented on docs/jetstream.md:
````diff
@@ -0,0 +1,213 @@
+            secretName: ""              # a secret containing the same keys as the "kubernetes.io/tls" type, if only a 
+                                        # `ca.crt` is present then we skip client verification but still ensure the 
+                                        # connection is encrypted with server verification.
+```
````
This is out of date, update docs at next opportunity

dan-j commented on docs/jetstream.md:
```diff
@@ -0,0 +1,213 @@
+- Channels across different namespaces cannot be starved of resources based on channels in other namespaces.
+- Channels across different namespaces can be configured with different credentials.
+
+#### Controller Dispatcher Configuration
```
Mark this section as a future enhancement, for now the config map must be named `config-nats` and there is no way to customise it

dan-j commented on pkg/channel/jetstream/dispatcher/reconciler.go:
```diff
@@ -0,0 +1,190 @@
+func (r *Reconciler) syncChannel(ctx context.Context, nc *v1alpha1.NatsJetStreamChannel, isLeader bool) pkgreconciler.Event {
+	logger := logging.FromContext(ctx).With(zap.Bool("is_leader", isLeader))
+
+	if !nc.Status.IsReady() {
+		logger.Debugw("NatsJetStreamChannel still not ready, short-circuiting the reconciler", zap.String("channel", nc.Name))
+		return nil
+	}
```
Thoughts... If we're not syncing the dispatcher with hostnames to listen for until after the channel is ready, then this means a sender could be sending messages to a channel before the dispatcher is ready to handle them.

If we're in a single replica dispatcher deployment, this can't happen since the channel won't be marked as ready until after the stream and consumer are created. So in this scenario there is no problem

This breaks down when there are multiple dispatcher replicas. The "leader" will mark the channel as ready, but other non-leader dispatcher pods may not have registered the hostnames yet. Since in k8s eyes this dispatcher pod is "ready", requests may be routed to it, but the receiver won't know how to handle it.

Not sure how to handle this:

1. Simple solution, but not ideal... Only allow single replicas
2. If a request is received on a host not registered, go and query the channel and manually call `ObserveKind` within the scope of the request. This will be slower, but at least it's handled, then the proper reconcile (which is probably only milliseconds away) will be a no-op.

Anybody any better ideas?

dan-j commented:
More updates to this now.. Everything seems working for cluster scoped channels, but namespace scoped channels results in an infinite loop of status updates, causing a reconcile (event though these returned reconciler events are meant to dequeue after 10 seconds, it's immediate). For some reason the dispatcher pods keep getting terminated and recreated, and my cluster just runs at 100% CPU and never becomes ready.

Any ideas why this might happen when deploying to another namespace and not knative-eventing? Everything works fine there.

dan-j commented:
> More updates to this now.. Everything seems working for cluster scoped channels, but namespace scoped channels results in an infinite loop of status updates, causing a reconcile (even though these returned reconciler events are meant to dequeue after 10 seconds, it's immediate). For some reason the dispatcher pods keep getting terminated and recreated, and my cluster just runs at 100% CPU and never becomes ready.
> 
> Any ideas why this might happen when deploying to another namespace and not knative-eventing? Everything works fine there.

Well, that was a rabbit hole. The dispatcher deployment had `ownerReferences` set which were referencing the controller pod in the `knative-eventing` namespace (different to `default`, which is where I was creating my "scoped" resources), which triggers a `OwnerRefInvalidNamespace` event on the kubelet. This clears the owner ref, and puts the deployment up for garbage collection. Controller is enqueued with the deletion, says, "create a new deployment", and repeat 🙃

dan-j commented:
The namespace-scoped issue has now been fixed. This should be ready for others to play around with now.

The only outstanding things are:

- [ ] Tests
- [ ] Dispatchers aren't removed when there are no more channels to be handled (but NATS resources are cleaned up)
- [ ] Owners refs on all namespace-scoped resources, this is only applied on the dispatcher deployment atm (see discussion point below)

Points for discussion:

- [ ] The jetstream-ch-controller cluster role is used as owner references on dispatcher deployments, the idea is so that if one uninstalls the controller, k8s will garbage collect everything else. You can't have a namespaced resource act as owner to another resource in a different namespace, so you can't use the `knative-eventing/jetstream-ch-controller` deployment as the owner. Is there a better resource to use?

cc @zhaojizhuang

dan-j commented on docs/local-development.md:
````diff
@@ -0,0 +1,46 @@
+Deploy eventing-natss to your cluster (ensuring your `kubectl` context is configured to your local cluster):
+
+```sh
+ko apply -f config/webhook
+ko apply -f config/jetstream
````
And you may also need the `--insecure-registry` flag set:

```sh
KO_DOCKER_REPO=k3d-ko.local:12345 ko apply --insecure-registry -f config/webhook
KO_DOCKER_REPO=k3d-ko.local:12345 ko apply --insecure-registry -f config/jetstream
```

zadus replied:
Hi @dan-j ! I've tried out this PR for the basic pingsource example works, following the instructions in local-development.md by applying the files in webhook and jetstream, seeing some cloudevents in the logs on the event-display service

I'm currently testing this locally using k3s running on Rancher Desktop

The files I deployed from the example folder are config-nats.yaml, nats.yaml and ping-demo-default.yaml

Will be reading through and understanding more on this implementation over the next few days

Foxman13 replied:
We would like to start developing with NATS JetStream and Knative. Is there anything my team can do help update/validate this PR, so it can be merged?

derekcollison commented:
Anything the Synadia team (NATS maintainers) can do to help?

Foxman13 commented:
@dan-j Please let me know if my team of engineers can help bring this PR along to acceptance. Also noticed an offer today to help from the NATS maintainers ^^^.

dan-j commented:
Hey all! Sorry for the radio silence, mad couple of months, life has got in the way of work a fair bit.

There are two things to look at regarding what's left to get this over the line: 

1. This comment: https://github.com/knative-sandbox/eventing-natss/pull/267#issuecomment-1044286149
2. And my own review of my latest push, which adds some comments for things I'd forgotten/not had time to tidy, or raises some points for discussion: https://github.com/knative-sandbox/eventing-natss/pull/267#pullrequestreview-882268453

If somebody wants to continue the PR then I'd be greatly appreciative. I'm happy to help people out if there are particular questions, either here or on the Knative slack (#eventing-questions). If somebody wants to address just one aspect of the outstanding items, perhaps contribute as a PR into my fork, then we can review and merge, then there's just one thing to merge here once everything's good for an RC/alpha release?

codegangsta commented:
@dan-j Hey there! Happy to get this over the line this week. Getting myself caught up on the context but should be able to handle those 2 items

Foxman13 commented:
@dan-j @codegangsta What's outstanding on this PR? Thank you

sjbodzo commented:
Would love to see this ship!

dan-j commented:
Hey all! Apologies for not being involved lately, lots of life events getting in the way.

Just pulling this down to familiarise myself where I got up to and will update the PR, either by code pushes or just comments. Still keen to push this through, but I started working on this for work and then priorities changed. We're close to needing this production-ready so I'm hopefully going to become more active again.

dan-j commented:
Massive thanks to @codegangsta to his contributions adding tests and some other fixes! I've merged dan-j/eventing-natss#1 into this branch ✅

I've address some of my previous comments with out of date documentation, just whilst I get back up to speed with things.

I'm only working Monday-Tuesday this week, so may not get much done this week but will try and look into the outstanding issues I mentioned on 18th Feb (https://github.com/knative-sandbox/eventing-natss/pull/267#issuecomment-1044286149)

Does anyone have any ideas on addressing this comment? Might be in a better headspace to think of something after a bit more time in the code, but last time I recall being a bit stumped. https://github.com/knative-sandbox/eventing-natss/pull/267#discussion_r806308945

dan-j commented:
I've just rebased this into three commits:
- my work up until @codegangsta added tests
- @codegangsta's work squashed into one commit (and kept the author)
- my changes since @codegangsta's work

Tests are passing and we're currently running this in a production environment now and things seem to be working ok.

I've noticed there have been recent changes in the current `v1alpha1/NatsJetStreamChannel` around tracing support. There may be other changes, I've not looked yet. I think it may be wise to release this as a `v1alpha2` version, however I don't know how we'd manage the upgrade path for users already using this. Another solution is to revisit the possibility of creating a new repo for JetStream and deprecating this one. It could be a big breaking change for anyone who wants to migrate to this version since the underlying logic of how JetStream is used is different too.

Thoughts?

dan-j commented:
> @dan-j are you referring to this #364? There is no API changes, only internal behavior. Introducing a v1alpha2 version won't solve your problem.
> 
> Are you doing tracing differently? What's the concern?

I was referring to the general fact that changes to JSM have been made, tracing was one of them, but I am also not 100% certain whether there have been other changes. It wasn't specifically aimed at tracing.

My point around a new version is twofold:

1. We've changed the CRD definition (although not in a breaking way, since we've only added fields)
2. We've changed the underlying implementation - which is technically a breaking change, you wouldn't be able to upgrade the controller in cluster and everything continue to "just work", even though the CRDs are backwards compatible.

Specifically around 2):

- You can't configure the NATS URL with the `DEFAULT_JETSTREAM_URL` environment variable anymore, you must have a `config-nats` ConfigMap.
- The stream names and subjects which events are published to are different. This may be out of scope for something we want to support but if users were relying on stream/subject names for other purposes outside of Knative, this would also break.

Finally, I'm not really too opinionated on this so I can be swayed either way... My argument for a new git repository as a whole is that NATS Streaming is deprecated by Synadia, we shouldn't be encouraging users to adopt it. Then this implementation of JetStream is massively different to the current version, and some people may be happy with what they've got and not want to upgrade. Merging this and deleting the current implementation leaves those people behind with no change of minor fixes (like tracing for example)

dan-j commented:
Additionally to my point of a new repo, it was said somewhere that we would just rename this repo to `eventing-nats` or `eventing-jetstream` once NATS Streaming was fully deprecated and removed, however that would be a breaking change with regards to image tags, go imports (although people shouldn't really be using this as a library). You could work around this but it's just more technical debt to manage

lionelvillard commented:
I would personally not worry too much about being backward-compatible at this development stage. However, to be sure, can you send an email to the knative-users mailing list warning them of the upcoming breaking changes?

astelmashenko commented:
Hey, 2 cents from our side, since we are using eventing-natss, right now we are migrating to jetstream.
This repo has latest release 1.4.0, it seems like it is not used widely. And if used with old versions of knative.

About @dan-j points:
> You can't configure the NATS URL with the DEFAULT_JETSTREAM_URL environment variable anymore, you must have a config-nats ConfigMap.
This is not a problem at all.

> The stream names and subjects which events are published to are different. This may be out of scope for something we want to support but if users were relying on stream/subject names for other purposes outside of Knative, this would also break.
Users of knative-eventing must not rely on internal implementation details, imho.

> Also to the previous point, events received but not delivered mid-upgrade would be lost due to stream name changes
This may be an issue, yes. No zero downtime release is available. But if you create a new repo these issues will be solved, right?

Again imho, it is more important to release this sooner than later then worrying about breaking changes of a project which was almost dead (even removed from release process).

dan-j commented:
I've been talking to @astelmashenko on Slack about the progress on this, just thought I'd share here.

We've been using this in production for ~2 months and have only come across one issue when using namespace-scoped channels. If you delete the last channel in the namespace, the `knative-eventing/jetstream-ch-controller` Pod deletes the `jetstream-ch-dispatcher` Pod from the scoped namespace, potentially before it's had chance to clean up the created Stream. In these instances I've had to manually delete the stream.

I'm trying my hardest to find time to address this PR but every time I think I'm going to have some time something new lands on my desk. I may have another colleague to put onto this, if nobody else can find the time I'll try and provide another update next Friday.

chris93111 commented:
Hi @dan-j i have open this issue https://github.com/knative-sandbox/eventing-natss/issues/381, i use the 1.3.5 release with new crd and sequence broker knative, the new dispatche accept only content-type: application/json ?

dan-j commented:
@chris93111 is this using the new implementation from this PR or the original one?

This implementation _should_ support any valid CloudEvent following the HTTP specification:

- Either binary encoded events, content-type may be anything and ce- specific headers on the request
- Or structured encoding where content-type is application/cloudevents+json etc.

Do you have any more info on the actual requests being sent?

chris93111 commented:
@dan-j i use the release 1.3.5 manifest in https://github.com/knative-sandbox/eventing-natss/releases/tag/knative-v1.3.5

try this tutorial https://knative.dev/docs/eventing/flows/sequence/sequence-with-broker-trigger/#create-the-trigger-targeting-the-sequence , it's work with 1.3.0 before new crd

Now the channel dispatcher display error failed to forward reply to http://$broker specversion: no specversion 

The event is not send to the broker

dan-j commented:
Sorry @chris93111, only just seen your reply.

Perhaps @astelmashenko may be able to comment? I haven't tested this new implementation using Brokers at all, but he did the port to the 1.3.5.

Can you attach all of the manifests and I can take a quick look?

dan-j commented:
I believe this is ready to be merged now. There's still a few outstanding issues with leftover Streams when running in namespace-scoped mode, but I can create new tickets. Would really like to get this merged in then I can stop having to rebase each time I look at it.

If there's anything anyone would like me to address before merging let me know

dan-j commented:
Hey @codegangsta! Have you signed the CLA agreement by any chance? There was a warning about it by the bot here: https://github.com/knative-sandbox/eventing-natss/pull/267#issuecomment-1250157153

astelmashenko commented:
> Sorry @chris93111, only just seen your reply.
> 
> Perhaps @astelmashenko may be able to comment? I haven't tested this new implementation using Brokers at all, but he did the port to the 1.3.5.
> 
> Can you attach all of the manifests and I can take a quick look?

Yes, I'd like to add a patch with tracing headers propagation (there is a diff inside zip):
[tracing_propagation.zip](https://github.com/knative-sandbox/eventing-natss/files/10309948/tracing_propagation.zip)

astelmashenko commented:
To test brokers you can update a config-map. Configure configmap name:`config-br-default-channel` ns:`knative-eventing`:
```yaml
data:
  channel-template-spec: |
    apiVersion: messaging.knative.dev/v1alpha1
    kind: NatsJetStreamChannel
    spec:
      stream:
        config:
          retention: Limits
          maxBytes: 1000000000
      consumerConfigTemplate:
        deliverPolicy: New
```

chris93111 commented:
> Sorry @chris93111, only just seen your reply.
> 
> Perhaps @astelmashenko may be able to comment? I haven't tested this new implementation using Brokers at all, but he did the port to the 1.3.5.
> 
> Can you attach all of the manifests and I can take a quick look?

Hi @dan-j you can found manifest in issue https://github.com/knative-sandbox/eventing-natss/issues/381 or documentation https://knative.dev/docs/eventing/flows/sequence/sequence-with-broker-trigger/

astelmashenko commented:
@dan-j , one more thing you may be interested in is this PR https://github.com/knative-sandbox/eventing-natss/pull/380/files
it adds jetstream resources into release (see https://github.com/knative-sandbox/eventing-natss/releases/tag/knative-v1.3.5 there will be two yaml files: eventing-natss.yaml and eventing-jsm.yaml)

dan-j commented:
@astelmashenko and everyone else... I've added the jetstream and webhook manifests to the release process and included the requested diff for propagating tracing headers

chris93111 commented:
> @astelmashenko and everyone else... I've added the jetstream and webhook manifests to the release process and included the requested diff for propagating tracing headers

Hi @dan-j di have you idea for issue with broker ?

codegangsta commented:
> Hey @codegangsta! Have you signed the CLA agreement by any chance? There was a warning about it by the bot here: [#267 (comment)](https://github.com/knative-sandbox/eventing-natss/pull/267#issuecomment-1250157153)

Signed!

astelmashenko commented:
@dan-j , is anything else left or preventing us from merging to main?

chris93111 commented:
Hi @dan-j @astelmashenko have you found any idea of this issue https://github.com/knative-sandbox/eventing-natss/issues/381 ?

astelmashenko commented:
@zadus , @Foxman13 , what do we need to make this PR merged?
We are using this PR port (to version 1.3) in production. We'd like to migrate to later knative versions, but there is no eventing-natss version which we can use with JetStream (for now it is only release-1.3 only).
thanks

dan-j commented:
Hey @chris93111, I'm afraid I haven't had time to look, however we're using this in production processing 1000's of messages a day with no issue. Has anyone else been able to reproduce?

We are using the namespace-scoped channels (which I don't believe you can do using brokers due to the template not supporting annotations?) with multiple namespace deployments, rather than cluster-scoped with single dispatcher in `knative-eventing`, but I can't see why that should cause any issues.

From my perspective this is ready to merge

chris93111 commented:
Hi @dan-j thanks for your response , are you try this procedure https://knative.dev/docs/eventing/flows/sequence/sequence-with-broker-trigger/ ?

astelmashenko commented:
/assign @zhaojizhuang , please review this PR, it seems to be ready.

astelmashenko commented:
/assign @zhaojizhuang

astelmashenko commented:
@dan-j , let's try to make progress here. I do not see all the checks, if there are all e2e tests are running and green and if other actions are green. Can we enable all the github actions and make it fully running build? I can help with broken steps (I need to ask access to you fork repo).
Also trying to understand what knative release version this PR is targeting is it release-1.5?

pierDipi commented:
/lgtm
/approve

I'd love to see https://github.com/knative-sandbox/eventing-natss/pull/391 completed and merged

astelmashenko commented:
@pierDipi,

> I'd love to see https://github.com/knative-sandbox/eventing-natss/pull/391 completed and merged
PR 391 is the same as PR 267, I just created it to check e2e tests and other checks.

Since PR 267 is merged we can now create the next release, it should be release-1.5 I guess. I can see it references `v0.34.1-0.20220928014117-bd37622589a4` which version knative-eventing/serving is it? Do not you know which version of knative is referenced in this PR (now main)?
